### PR TITLE
Axes.draw w/o getBoundingClientRect + many axis automargin fixes

### PR DESCRIPTION
**Pull request by etpinard**
This PR resolves https://github.com/plotly/plotly.js/issues/1988, fixes https://github.com/plotly/plotly.js/issues/2434, improves https://github.com/plotly/plotly.js/issues/298 and should help us towards https://github.com/plotly/plotly.js/issues/2704

Before this PR, `ax._boundingBox` was computed in `Axes.drawOne` using a raw (and costly) `getBoundingClientRect` call. In turn, `ax._boundingBox` was used to: (1) position cartesian spikes (2) position range sliders (3) compute the axis auto-margin push value. Meaning that for most graphs, that `getBoundingClientRect` was called for no reason. Commit https://github.com/plotly/plotly.js/commit/d2650fe30a38a4997774acc3d66f97de44b0d989 introduces a better of doing things, see commit description for the details. Commit https://github.com/plotly/plotly.js/commit/2e7b3dd6143ac9fddbc200dcc883f08c96c4a175 essentially removes the need for computing a bounding box to render cartesian spikes.

Elsewhere in this PR, 

- the first two commits are clean-ups commits, 
- the next two are fixups of things I noticed while working in auto-margin land https://github.com/plotly/plotly.js/issues/2704
- https://github.com/plotly/plotly.js/commit/d54ec945304dd14bfe565d2aa5c91f030ea160c4 and https://github.com/plotly/plotly.js/commit/a7e8eaf62f1dbfa1093ae12db1b0d1869a7006da presents a few axis auto-margin bugs I discovered
- these get fixed in https://github.com/plotly/plotly.js/commit/daa88e04ee18cc5bdaaa39897215526f5cca92d2, https://github.com/plotly/plotly.js/commit/40f1b24167e8d80e1a242930aa45d90bc71b4954 and https://github.com/plotly/plotly.js/commit/bb127af41e57368ed319af29241c4802093107b0

cc @plotly/plotly_js

### CODE REVIEW

**Comment by etpinard** [comment]
Unfortunately, switching from using raw `getBoundingClientRect` to `Drawing.bBox` lead to few baseline changes. Please refer to each of the commit description for explanations.

File: src/plots/cartesian/axes.js [hunk 1]
```diff
@@ -2052,27 +2024,45 @@ function getDividerVals(ax, vals) {
...
+        ax._selections[cls].each(function() {
+            var thisLabel = selectTickLabel(this);
+            // Use parent node <g.(x|y)tick>, to make Drawing.bBox
+            // retrieve a bbox computed with transform info
+            //
+            // To improve perf, it would be nice to use `thisLabel.node()`
+            // (like in fixLabelOverlaps) instead and use Axes.getPxPosition
+            // together with the makeLabelFns outputs and `tickangle`
+            // to compute one bbox per (tick value x tick style)
+            var bb = Drawing.bBox(thisLabel.node().parentNode);
```
Comment (etpinard) [inline src/plots/cartesian/axes.js]: We should probably copy this comment over to https://github.com/plotly/plotly.js/issues/298 - after this PR gets merged.

File: src/plots/cartesian/axes.js [hunk 2]
```diff
@@ -1636,6 +1636,24 @@ axes.draw = function(gd, arg, opts) {
...
+ *
+ * Depends on:
+ * - ax._mainSubplot (from linkSubplots)
+ * - ax._mainAxis
+ * - ax._anchorAxis
+ * - ax._subplotsWith
+ * - ax._counterDomainMin, ax._counterDomainMax (optionally, from linkSubplots)
+ * - ax._mainLinePosition (from lsInner)
+ * - ax._mainMirrorPosition
+ * - ax._linepositions
```
Comment (etpinard) [inline src/plots/cartesian/axes.js]: Computing and stashing `_mainLinePosition`, `_mainMirrorPosition` and `_linepositions` in `lsInner` is less than ideal. This makes `Axes.drawOne` "depend" on `lsInner` i.e. `lsInner` must get called before `Axes.drawOne` on first draw and `'ticks'` edits.

---

Now in relation to https://github.com/plotly/plotly.js/issues/2704, if we choose to add a "pure margin computation" step e.g. called `Axes.pushMargin`, we'll need to un-coupled `drawOne` and `lsInner`. After this PR, `Axes.pushMargin` for a single axis could look like:

```js
axes.pushMargin = function(gd, ax) {
  var seq = [];
  
  /* compute mainLinePosition */

  seq.push(function() {
    return drawLabels(gd, ax, {
      vals: axes.calcTicks(ax),
      layer: /* some dummy layer in the tester div */,
      transFn: axes.makeTransFn(ax),
      labelFns: axes.makeLabelFns(ax, mainLinePosition)
    })
  })

  /* N.B. we don't need to draw ticks, grid lines nor axis lines
     to compute the margin push values */

  seq.push(function() {
    /* compute the margin push values */ 

    Plots.autoMargin(gd, /* */) 
  })
}
```

or maybe we could have a set of getter `Axes` helpers so we won't have to stash the values.

File: src/plots/cartesian/axes.js [hunk 3]
```diff
@@ -1813,168 +1843,110 @@ axes.drawOne = function(gd, ax, opts) {
...
-                    width: 0
-                };
-            }
-        }
-    }
-
     var hasRangeSlider = Registry.getComponentMethod('rangeslider', 'isVisible')(ax);
 
-    function doAutoMargins() {
+    seq.push(function() {
```
Comment (etpinard) [inline src/plots/cartesian/axes.js]: This diff may be a little hard to read, but essentially this commits :hocho: the `calcBoundingBox` step and makes the margin push value computation step this:

https://github.com/plotly/plotly.js/blob/d2650fe30a38a4997774acc3d66f97de44b0d989/src/plots/cartesian/axes.js#L1864-L1946

**Comment by antoinerg** [comment]
Pretty solid PR @etpinard :muscle: It's well commented and the commits are well organized. 

I'm tempted to give you a dancer but I feel like someone else should also take a look at it considering it is touching cartesian axes. @alexcjohnson would be an ideal reviewer.

File: src/plots/cartesian/axes.js [hunk 4]
```diff
@@ -2052,27 +2024,45 @@ function getDividerVals(ax, vals) {
...
+            // To improve perf, it would be nice to use `thisLabel.node()`
+            // (like in fixLabelOverlaps) instead and use Axes.getPxPosition
+            // together with the makeLabelFns outputs and `tickangle`
+            // to compute one bbox per (tick value x tick style)
+            var bb = Drawing.bBox(thisLabel.node().parentNode);
+            top = Math.min(top, bb.top);
+            bottom = Math.max(bottom, bb.bottom);
+            left = Math.min(left, bb.left);
+            right = Math.max(right, bb.right);
+        });
```
[the remark this answers is not in the document]
Comment (etpinard) [reply src/plots/cartesian/axes.js]: I don't think `getBoundingClientRect` can even return an `Infinity`.

[the remark this answers is not in the document]
Comment (archmoj) [reply src/plots/cartesian/axes.js]: OK. Nothing to worry there.

**Comment by archmoj** [comment]
@antoinerg Looking good to me as well.
@etpinard :dancer: :dancer:

**Comment by archmoj** [comment]
Thanks @alexcjohnson for the review.

**Pull request merged**